src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+  goto_convertt converter(symbol_table, message_handler);
+
+  // HEAD: // Loop body block
+  //   ... eval guard ...
+  //   IF (!guard) GOTO EXIT;
+  //   ... loop body ...
+  //   // instrumentation
+  //   entered_loop = true
+  //   // Jump back to the step case if the loop can run at least once
+  //   IF (in_base_case) GOTO STEP;
+  //   ASSERT(<loop_invariant>);
+  //    new_variant = <loop_decreases>;
+  //   ASSERT(new_variant < old_variant);
+  //   ASSUME(false);
+
+  goto_programt pre_loop_latch_instrs;
+
+  {
+    // Record that we entered the loop with `entered_loop = true`
+    pre_loop_latch_instrs.add(
+      goto_programt::make_assignment(entered_loop, true_exprt{}));
+  }
+
+  {
+    // Jump back to the step case to havoc the write set, assume the invariant,
+    // and execute an arbitrary iteration.
+    pre_loop_latch_instrs.add(goto_programt::make_goto(
+      step_case_target, in_base_case, loop_head_location));
+  }
+
+  {
+    // Because of the unconditional jump above the following code is only
+    // reachable in the step case. Generate the inductive invariant check
+    // `ASSERT(invariant)`.
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_invariant_step");
+    check_location.set_comment(
+      "Check invariant after step for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    code_assertt assertion{invariant};
+    assertion.add_source_location() = check_location;
+    converter.goto_convert(assertion, pre_loop_latch_instrs, language_mode);
+  }
+
+  {
+    // Generate assignments to store the multidimensional decreases clause's
+    // value after one iteration of the loop
+    if(!decreases_clauses.empty())
+    {
+      for(size_t i = 0; i < new_decreases_vars.size(); i++)
+      {
+        code_assignt new_decreases_assignment{
+          new_decreases_vars[i], decreases_clauses[i]};
+        new_decreases_assignment.add_source_location() = loop_head_location;
+        converter.goto_convert(
+          new_decreases_assignment, pre_loop_latch_instrs, language_mode);
+      }
+
+      // Generate assertion that the multidimensional decreases clause's value
+      // after the loop is lexicographically smaller than its initial value.
+      source_locationt check_location(loop_head_location);
+      check_location.set_property_class("loop_decreases");
+      check_location.set_comment(
+        "Check variant decreases after step for loop " +
+        id2string(check_location.get_function()) + "." +
+        std::to_string(cbmc_loop_id));
+      code_assertt monotonic_decreasing_assertion{
+        generate_lexicographic_less_than_check(
+          new_decreases_vars, old_decreases_vars)};
+      monotonic_decreasing_assertion.add_source_location() = check_location;
+      converter.goto_convert(
+        monotonic_decreasing_assertion, pre_loop_latch_instrs, language_mode);
```

Comment:
Do you really need to use `goto_convert` here? Couldn't you directly add an `assert` instruction?

Reply:
We don't need `goto_convert` here. Both `new_decreases_vars` and `old_decreases_vars` are symbol expressions.